File: overlay/src/contracts/presentation.ts
export type SourceStatus = 'online' | 'offline' | 'unknown'
export type DisplayFieldKind =
  | 'text'
  | 'integer'
  | 'duration_ms'
  | 'timestamp'
  | 'latency_ms'
  | 'coordinates'
  | 'status'
export type DisplayTone = 'normal' | 'warning' | 'danger' | 'muted'

export interface PresentationIdentity {
  display_name: string
  account_name?: string
  level?: number
}

export interface PresentationMap {
  x: number
  y: number
  projection: string
  tile_set: string
  tile_url: string
}

interface DisplayFieldBase<K extends DisplayFieldKind> {
  id: string
  label: string
  kind: K
  tone: DisplayTone
}

interface AvailableField<K extends DisplayFieldKind, V> extends DisplayFieldBase<K> {
  available: true
  value: V
  progress?: number
}

export type TextDisplayField = AvailableField<'text', string>
export type IntegerDisplayField = AvailableField<'integer', number>
export type DurationDisplayField = AvailableField<'duration_ms', number>
export type TimestampDisplayField = AvailableField<'timestamp', string>
export type LatencyDisplayField = AvailableField<'latency_ms', number>
export type CoordinatesDisplayField = AvailableField<
  'coordinates',
  { x: number; y: number }
>
export type StatusDisplayField = AvailableField<'status', string>

export type AvailableDisplayField =
  | TextDisplayField
  | IntegerDisplayField
  | DurationDisplayField
  | TimestampDisplayField
  | LatencyDisplayField
  | CoordinatesDisplayField
  | StatusDisplayField

export interface UnavailableDisplayField extends DisplayFieldBase<DisplayFieldKind> {
  available: false
}

export type DisplayField = AvailableDisplayField | UnavailableDisplayField

export interface Presentation {
  schema: 'overlay.presentation/v1'
  game_id: string
  user_id: string
  observed_at: string
  fresh_until: string
  source_status: SourceStatus
  identity: PresentationIdentity
  map?: PresentationMap
  fields: DisplayField[]
}

const RFC3339 = /^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})(?:\.\d+)?(?:Z|([+-])(\d{2}):(\d{2}))$/
const SAFE_FIELD_ID = /^[a-z0-9][a-z0-9._-]{0,95}$/

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
}

function record(value: unknown, path: string): Record<string, unknown> {
  if (!isRecord(value)) {
    throw new Error(`${path} must be an object`)
  }
  return value
}

function exactKeys(
  value: Record<string, unknown>,
  required: readonly string[],
  optional: readonly string[],
  path: string,
): void {
  const allowed = new Set([...required, ...optional])
  for (const key of Object.keys(value)) {
    if (!allowed.has(key)) throw new Error(`${path}.${key} is unknown`)
  }
  for (const key of required) {
    if (!Object.hasOwn(value, key)) throw new Error(`${path}.${key} is required`)
  }
}

function nonEmptyString(value: unknown, path: string): string {
  if (typeof value !== 'string' || value.trim() === '') {
    throw new Error(`${path} must be a non-empty string`)
  }
  return value
}

function finiteNumber(value: unknown, path: string): number {
  if (typeof value !== 'number' || !Number.isFinite(value)) {
    throw new Error(`${path} must be a finite number`)
  }
  return value
}

function safeInteger(value: unknown, path: string): number {
  const parsed = finiteNumber(value, path)
  if (!Number.isSafeInteger(parsed)) throw new Error(`${path} must be a safe integer`)
  return parsed
}

function rfc3339(value: unknown, path: string): string {
  const parsed = nonEmptyString(value, path)
  const match = RFC3339.exec(parsed)
  if (!match || Number.isNaN(Date.parse(parsed))) {
    throw new Error(`${path} must be an RFC3339 timestamp`)
  }

  const [, year, month, day, hour, minute, second, , offsetHour, offsetMinute] = match
  const yearNumber = Number(year)
  const monthNumber = Number(month)
  const dayNumber = Number(day)
  const daysInMonth = new Date(Date.UTC(yearNumber, monthNumber, 0)).getUTCDate()
  if (
    monthNumber < 1 || monthNumber > 12 || dayNumber < 1 || dayNumber > daysInMonth ||
    Number(hour) > 23 || Number(minute) > 59 || Number(second) > 59 ||
    Number(offsetHour ?? 0) > 23 || Number(offsetMinute ?? 0) > 59
  ) {
    throw new Error(`${path} must be an RFC3339 timestamp`)
  }
  return parsed
}

function displayKind(value: unknown, path: string): DisplayFieldKind {
  switch (value) {
    case 'text':
    case 'integer':
    case 'duration_ms':
    case 'timestamp':
    case 'latency_ms':
    case 'coordinates':
    case 'status':
      return value
    default:
      throw new Error(`${path} is unsupported`)
  }
}

function displayTone(value: unknown, path: string): DisplayTone {
  switch (value) {
    case 'normal':
    case 'warning':
    case 'danger':
    case 'muted':
      return value
    default:
      throw new Error(`${path} is unsupported`)
  }
}

function sourceStatus(value: unknown): SourceStatus {
  switch (value) {
    case 'online':
    case 'offline':
    case 'unknown':
      return value
    default:
      throw new Error('source_status is unknown')
  }
}

function progress(value: unknown, path: string): number {
  const parsed = finiteNumber(value, path)
  if (parsed < 0 || parsed > 1) throw new Error(`${path} must be between 0 and 1`)
  return parsed
}

function parseIdentity(value: unknown): PresentationIdentity {
  const raw = record(value, 'identity')
  exactKeys(raw, ['display_name'], ['account_name', 'level'], 'identity')
  const identity: PresentationIdentity = {
    display_name: nonEmptyString(raw.display_name, 'identity.display_name'),
  }
  if (Object.hasOwn(raw, 'account_name')) {
    identity.account_name = nonEmptyString(raw.account_name, 'identity.account_name')
  }
  if (Object.hasOwn(raw, 'level')) identity.level = safeInteger(raw.level, 'identity.level')
  return identity
}

function parseMap(value: unknown): PresentationMap {
  const raw = record(value, 'map')
  exactKeys(raw, ['x', 'y', 'projection', 'tile_set', 'tile_url'], [], 'map')
  return {
    x: finiteNumber(raw.x, 'map.x'),
    y: finiteNumber(raw.y, 'map.y'),
    projection: nonEmptyString(raw.projection, 'map.projection'),
    tile_set: nonEmptyString(raw.tile_set, 'map.tile_set'),
    tile_url: nonEmptyString(raw.tile_url, 'map.tile_url'),
  }
}

function parseCoordinates(value: unknown, path: string): { x: number; y: number } {
  const raw = record(value, path)
  exactKeys(raw, ['x', 'y'], [], path)
  return {
    x: finiteNumber(raw.x, `${path}.x`),
    y: finiteNumber(raw.y, `${path}.y`),
  }
}

function parseField(value: unknown, index: number): DisplayField {
  const path = `fields[${index}]`
  const raw = record(value, path)
  exactKeys(raw, ['id', 'label', 'kind', 'available', 'tone'], ['value', 'progress'], path)

  const id = nonEmptyString(raw.id, `${path}.id`)
  if (!SAFE_FIELD_ID.test(id)) throw new Error(`${path}.id is unsafe`)
  const label = nonEmptyString(raw.label, `${path}.label`)
  const kind = displayKind(raw.kind, `${path}.kind`)
  const tone = displayTone(raw.tone, `${path}.tone`)
  if (typeof raw.available !== 'boolean') throw new Error(`${path}.available must be boolean`)

  if (!raw.available) {
    if (Object.hasOwn(raw, 'value') || Object.hasOwn(raw, 'progress')) {
      throw new Error(`${path} is unavailable but carries value or progress`)
    }
    return { id, label, kind, available: false, tone }
  }
  if (!Object.hasOwn(raw, 'value')) throw new Error(`${path}.value is required`)
  const parsedProgress = Object.hasOwn(raw, 'progress')
    ? progress(raw.progress, `${path}.progress`)
    : undefined
  const common = { id, label, available: true as const, tone }
  const withProgress = parsedProgress === undefined ? {} : { progress: parsedProgress }

  switch (kind) {
    case 'text':
    case 'status': {
      if (typeof raw.value !== 'string') throw new Error(`${path}.value must be a string`)
      return { ...common, kind, value: raw.value, ...withProgress }
    }
    case 'integer':
      return { ...common, kind, value: safeInteger(raw.value, `${path}.value`), ...withProgress }
    case 'duration_ms':
    case 'latency_ms':
      return { ...common, kind, value: finiteNumber(raw.value, `${path}.value`), ...withProgress }
    case 'timestamp':
      return { ...common, kind, value: rfc3339(raw.value, `${path}.value`), ...withProgress }
    case 'coordinates':
      return {
        ...common,
        kind,
        value: parseCoordinates(raw.value, `${path}.value`),
        ...withProgress,
      }
  }
}

function parseFields(value: unknown): DisplayField[] {
  if (!Array.isArray(value)) throw new Error('fields must be an array')
  const fields = value.map(parseField)
  const ids = new Set<string>()
  for (const field of fields) {
    if (ids.has(field.id)) throw new Error(`field ID ${field.id} is duplicate`)
    ids.add(field.id)
  }
  return fields
}

export function parsePresentation(value: unknown): Presentation {
  const raw = record(value, 'presentation')
  exactKeys(
    raw,
    [
      'schema',
      'game_id',
      'user_id',
      'observed_at',
      'fresh_until',
      'source_status',
      'identity',
      'fields',
    ],
    ['map'],
    'presentation',
  )
  if (raw.schema !== 'overlay.presentation/v1') {
    throw new Error('unsupported presentation schema')
  }

  const parsed: Presentation = {
    schema: 'overlay.presentation/v1',
    game_id: nonEmptyString(raw.game_id, 'game_id'),
    user_id: nonEmptyString(raw.user_id, 'user_id'),
    observed_at: rfc3339(raw.observed_at, 'observed_at'),
    fresh_until: rfc3339(raw.fresh_until, 'fresh_until'),
    source_status: sourceStatus(raw.source_status),
    identity: parseIdentity(raw.identity),
    fields: parseFields(raw.fields),
  }
  if (Object.hasOwn(raw, 'map')) parsed.map = parseMap(raw.map)
  return parsed
}

File: internal/overlay/presentation.go
package overlay

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"math"
	"strconv"
	"time"
)

const PresentationSchemaV1 = "overlay.presentation/v1"

type Presentation struct {
	Schema       string         `json:"schema"`
	GameID       string         `json:"game_id"`
	UserID       string         `json:"user_id"`
	ObservedAt   time.Time      `json:"observed_at"`
	FreshUntil   time.Time      `json:"fresh_until"`
	SourceStatus string         `json:"source_status"`
	Identity     Identity       `json:"identity"`
	Map          *MapPosition   `json:"map,omitempty"`
	Fields       []DisplayField `json:"fields"`
}

type DisplayField struct {
	ID        string          `json:"id"`
	Label     string          `json:"label"`
	Kind      string          `json:"kind"`
	Available bool            `json:"available"`
	Value     json.RawMessage `json:"value,omitempty"`
	Tone      string          `json:"tone"`
	Progress  *float64        `json:"progress,omitempty"`
}

type displayCoordinates struct {
	X float64 `json:"x"`
	Y float64 `json:"y"`
}

func StringDisplayField(id, label, kind, value, tone string, progress *float64) DisplayField {
	if kind != "text" && kind != "status" {
		panic("overlay: string display field requires text or status kind")
	}
	return availableDisplayField(id, label, kind, value, tone, progress)
}

func NumberDisplayField(id, label, kind string, value float64, tone string, progress *float64) DisplayField {
	if kind != "integer" && kind != "duration_ms" && kind != "latency_ms" {
		panic("overlay: number display field requires a numeric kind")
	}
	if !finite(value) || kind == "integer" && math.Trunc(value) != value {
		panic("overlay: number display field requires a valid numeric value")
	}
	return availableDisplayField(id, label, kind, value, tone, progress)
}

func TimestampDisplayField(id, label string, value time.Time, tone string) DisplayField {
	return availableDisplayField(id, label, "timestamp", value.Format(time.RFC3339Nano), tone, nil)
}

func CoordinatesDisplayField(id, label string, x, y float64, tone string) DisplayField {
	if !finite(x) || !finite(y) {
		panic("overlay: coordinates display field requires finite values")
	}
	return availableDisplayField(id, label, "coordinates", displayCoordinates{X: x, Y: y}, tone, nil)
}

func UnavailableDisplayField(id, label, kind, tone string) DisplayField {
	return DisplayField{ID: id, Label: label, Kind: kind, Tone: tone}
}

func availableDisplayField(id, label, kind string, value any, tone string, progress *float64) DisplayField {
	encoded, err := json.Marshal(value)
	if err != nil {
		panic(fmt.Sprintf("overlay: marshal display field value: %v", err))
	}
	return DisplayField{ID: id, Label: label, Kind: kind, Available: true, Value: encoded, Tone: tone, Progress: progress}
}

func ValidatePresentation(presentation Presentation) error {
	if presentation.Schema != PresentationSchemaV1 {
		return fmt.Errorf("presentation schema %q is unsupported", presentation.Schema)
	}
	seen := make(map[string]struct{}, len(presentation.Fields))
	for index, field := range presentation.Fields {
		if !safeDisplayFieldID(field.ID) {
			return fmt.Errorf("presentation field %d has unsafe ID %q", index, field.ID)
		}
		if _, exists := seen[field.ID]; exists {
			return fmt.Errorf("presentation field ID %q is duplicate", field.ID)
		}
		seen[field.ID] = struct{}{}
		if !supportedDisplayKind(field.Kind) {
			return fmt.Errorf("presentation field %q has unsupported kind %q", field.ID, field.Kind)
		}
		if !supportedDisplayTone(field.Tone) {
			return fmt.Errorf("presentation field %q has unsupported tone %q", field.ID, field.Tone)
		}
		if !field.Available {
			if len(field.Value) != 0 || field.Progress != nil {
				return fmt.Errorf("presentation field %q is unavailable but carries value or progress", field.ID)
			}
			continue
		}
		if len(field.Value) == 0 {
			return fmt.Errorf("presentation field %q is available but has no value", field.ID)
		}
		if field.Progress != nil && (!finite(*field.Progress) || *field.Progress < 0 || *field.Progress > 1) {
			return fmt.Errorf("presentation field %q has invalid progress", field.ID)
		}
		if err := validateDisplayValue(field.Kind, field.Value); err != nil {
			return fmt.Errorf("presentation field %q has invalid %s value: %w", field.ID, field.Kind, err)
		}
	}
	return nil
}

func safeDisplayFieldID(id string) bool {
	if len(id) == 0 || len(id) > 96 || !lowerAlphaNumeric(id[0]) {
		return false
	}
	for i := 1; i < len(id); i++ {
		if !lowerAlphaNumeric(id[i]) && id[i] != '.' && id[i] != '_' && id[i] != '-' {
			return false
		}
	}
	return true
}

func lowerAlphaNumeric(value byte) bool {
	return value >= 'a' && value <= 'z' || value >= '0' && value <= '9'
}

func supportedDisplayKind(kind string) bool {
	switch kind {
	case "text", "integer", "duration_ms", "timestamp", "latency_ms", "coordinates", "status":
		return true
	default:
		return false
	}
}

func supportedDisplayTone(tone string) bool {
	switch tone {
	case "normal", "warning", "danger", "muted":
		return true
	default:
		return false
	}
}

func validateDisplayValue(kind string, raw json.RawMessage) error {
	switch kind {
	case "text", "status":
		var value any
		if err := decodeSingleJSON(raw, &value); err != nil {
			return err
		}
		if _, ok := value.(string); !ok {
			return errors.New("must be a string")
		}
		return nil
	case "integer", "duration_ms", "latency_ms":
		number, err := decodeJSONNumber(raw)
		if err != nil {
			return err
		}
		value, err := number.Float64()
		if err != nil || !finite(value) {
			return errors.New("must be a finite number")
		}
		if kind == "integer" && math.Trunc(value) != value {
			return errors.New("must be an integer")
		}
		return nil
	case "timestamp":
		var value string
		if err := decodeSingleJSON(raw, &value); err != nil {
			return err
		}
		if _, err := time.Parse(time.RFC3339Nano, value); err != nil {
			return err
		}
		return nil
	case "coordinates":
		var object map[string]json.RawMessage
		if err := decodeSingleJSON(raw, &object); err != nil {
			return err
		}
		if len(object) != 2 {
			return errors.New("must contain exactly x and y")
		}
		for _, key := range []string{"x", "y"} {
			coordinate, ok := object[key]
			if !ok {
				return fmt.Errorf("missing %s", key)
			}
			number, err := decodeJSONNumber(coordinate)
			if err != nil {
				return fmt.Errorf("%s: %w", key, err)
			}
			value, err := number.Float64()
			if err != nil || !finite(value) {
				return fmt.Errorf("%s must be finite", key)
			}
		}
		return nil
	default:
		return errors.New("unsupported kind")
	}
}

func decodeJSONNumber(raw json.RawMessage) (json.Number, error) {
	var value any
	if err := decodeSingleJSON(raw, &value); err != nil {
		return "", err
	}
	number, ok := value.(json.Number)
	if !ok {
		return "", errors.New("must be a number")
	}
	if _, err := strconv.ParseFloat(number.String(), 64); err != nil {
		return "", errors.New("must be a number")
	}
	return number, nil
}

func decodeSingleJSON(raw json.RawMessage, destination any) error {
	decoder := json.NewDecoder(bytes.NewReader(raw))
	decoder.UseNumber()
	if err := decoder.Decode(destination); err != nil {
		return err
	}
	if err := decoder.Decode(&struct{}{}); !errors.Is(err, io.EOF) {
		return errors.New("must contain one JSON value")
	}
	return nil
}
